Sources/Tentacle/Client.swift

```diff
@@ -245,19 +228,22 @@ public final class Client {
         _ request: Request<[Resource]>,
         page: UInt? = 1,
         perPage: UInt? = 30
-    ) -> SignalProducer<(Response, [Resource]), Error> where Resource.DecodedType == Resource {
+    ) -> SignalProducer<(Response, [Resource]), Error> {
         let nextPage = (page ?? 1) + 1
+
         return execute(request, page: page, perPage: perPage)
-            .attemptMap { (response: Response, json: JSON) in
-                return decode(json)
-                    .map { resource in
-                        (response, resource)
-                    }
+            .attemptMap { response, data -> Result<(Response, [Resource]), Client.Error> in
+                return decodeList(data)
+                    .map { (response, $0) }
                     .mapError(Error.jsonDecodingError)
             }
-            .flatMap(.concat) { response, json -> SignalProducer<(Response, [Resource]), Error> in
-                return SignalProducer(value: (response, json))
-                    .concat(response.links["next"] == nil ? SignalProducer.empty : self.execute(request, page: nextPage, perPage: perPage))
+            .flatMap(.concat) { response, data -> SignalProducer<(Response, [Resource]), Error> in
+                let current = SignalProducer<(Response, [Resource]), Error>(value: (response, data))
+                guard let _ = response.links["next"] else {
+                    return current.concat(.empty)
```

Comment:
`.concat(.empty)` is effectively a noop. This can just be `return current`.

Reply:
Fixed!